### PR TITLE
Fix list hashcode

### PR DESCRIPTION
**Pull request by isaksky**

### CODE REVIEW

File: Xledger.Sql/Collections/ImmList.cs [hunk 1]
```diff
@@ -119,10 +119,8 @@ public bool Equals(ImmList<T> other) {
 
         public override int GetHashCode() {
             var h = 17;
-            foreach (var e in data) {
-                if (e == null) {
-                    h = h * 23 + e.GetHashCode();
-                }
+            for (int i = 0; i < data.Length; i++) {
+                h = h * 23 + data[i]?.GetHashCode() ?? 0;
```
Comment (oconnor0) [inline Xledger.Sql/Collections/ImmList.cs]:
```suggestion
                h = h * 23 + (data[i]?.GetHashCode() ?? 0);
```

The precedence of `??` is low enough that it ends up returning 0 and "erasing" h if there's a null in the list.

File: Xledger.Sql/Collections/ImmList.cs [hunk 2]
```diff
@@ -89,8 +89,8 @@ internal static IReadOnlyList<U> SelectList<U>(IList<T> list, Func<T, U> f) {
         public int Count => data.Length;
 
         public IEnumerator<T> GetEnumerator() {
-            foreach (var e in data) {
-                yield return e;
+            for (int i = 0; i < data.Length; i++) {
+                yield return data[i];
```
Comment (oconnor0) [inline Xledger.Sql/Collections/ImmList.cs]:
```suggestion
            return ((IEnumerable<T>)data).GetEnumerator();
```

Not sure if this is better.

File: Xledger.Sql/Collections/ImmList.cs (hunk 1 above)
Comment (isaksky) [reply Xledger.Sql/Collections/ImmList.cs]: Oh, scary

File: Xledger.Sql/Collections/ImmList.cs (hunk 2 above)
Comment (isaksky) [reply Xledger.Sql/Collections/ImmList.cs]: Probably is - fixed

**Review by oconnor0 — APPROVED** [review]

**Pull request merged**